pkg/client/contract.go

```diff
@@ -108,6 +133,42 @@ func (contract *Contract) Submit(transactionName string, options ...ProposalOpti
 	return result, nil
 }
 
+// SubmitWithContext submit a transaction to the ledger in the scope of a specific Context and return its result only after it has been committed to the ledger. This method
+// provides greater control over the transaction proposal content and the endorsing peers on which it is evaluated. This
+// allows transaction functions to be submitted where the proposal must include transient data, or that will access
+// ledger data with key-based endorsement policies.
+//
+// This method may return different error types depending on the point in the transaction invocation that a failure
+// occurs. The error can be inspected with errors.Is or errors.As.
+func (contract *Contract) SubmitWithContext(ctx context.Context, transactionName string, options ...ProposalOption) ([]byte, error) {
+	proposal, err := contract.NewProposal(transactionName, options...)
+	if err != nil {
+		return nil, err
+	}
+
+	transaction, err := proposal.EndorseWithContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	commit, err := transaction.SubmitWithContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	status, err := commit.StatusWithContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	if !status.Successful {
+		return nil, fmt.Errorf("transaction %s rejected with validation code %d (%s)", status.TransactionID, int32(status.Code), peer.TxValidationCode_name[int32(status.Code)])
+	}
+
+	result := transaction.Result()
```

Comment:
The `SubmitAsyncWithContext()` already does most of the work and can be reused rather than duplicated.

The `Submit()` and `SubmitAsync()` implementations intentionally return the `result` along with any error. Depending on where the a occurs it may be possible to have a non-nil result and a non-nil error, and this is actually valid information the client might want to use. This function should behave the same way. If there isn't one already, there should probably be a unit test to specify this behaviour, although that would be my omission from the initial implementation so it doesn't need to be done in this PR.

```suggestion
	result, commit, err := contract.SubmitAsyncWithContext(ctx, transactionName, options...)
	if err != nil {
		return result, err
	}
	status, err := commit.StatusWithContext(ctx)
	if err != nil {
		return result, err
	}
	if !status.Successful {
		return nil, newCommitError(status.TransactionID, status.Code)
	}

```